### PR TITLE
Adding Together as safety API provider

### PR DESCRIPTION
**Pull request by yogishbaliga**
For testing, brought up the local instance of the llama stack and ran few safety queries with input prompt to check. Verified that the output looks as expected.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -8,3 +8,5 @@ python-dotenv
 pydantic
 requests
 termcolor
+pytest
```
Comment (ashwinb) [inline requirements.txt:11]: Can we revert these here and add a `requirements.dev.txt` perhaps where these are added? The `llama-stack` package is used for some lightweight CLI commands (predominantly `llama model download`) and we don't want those folks to need unnecessary dependencies.

File: llama_stack/providers/registry/safety.py [hunk 2]
```diff
@@ -30,6 +31,20 @@ def available_providers() -> List[ProviderSpec]:
                 pip_packages=[],
                 module="llama_stack.providers.adapters.safety.sample",
                 config_class="llama_stack.providers.adapters.safety.sample.SampleConfig",
+            )
+        ),
+
```
Comment (ashwinb) [inline llama_stack/providers/registry/safety.py:36]: nit: remove newline?

File: llama_stack/providers/adapters/safety/together/safety.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+class TogetherSafetyImpl(Safety):
+    def __init__(self, config: TogetherSafetyConfig) -> None:
+        self.config = config
+
+    async def initialize(self) -> None:
+        pass
+
+    async def run_shield(
+            self,
```
Comment (ashwinb) [inline llama_stack/providers/adapters/safety/together/safety.py:28]: formatting looks off -- could you install `pre-commit` and run it once?

File: llama_stack/providers/adapters/safety/together/safety.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        together_api_key = self.config.api_key
+        # @TODO error out if together_api_key is missing in the header
+        if provider_data is not None:
+            if isinstance(provider_data, TogetherHeaderInfo):
+                together_api_key = provider_data.together_api_key
+
+        # messages can have role assistant or user
+        api_messages = []
+        for message in messages:
+            if type(message) is UserMessage:
```
Comment (ashwinb) [inline llama_stack/providers/adapters/safety/together/safety.py:48]: can you do `if message.role == Role.user.value`? A bit more predictable that is

File: llama_stack/providers/adapters/safety/together/safety.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                api_messages.append({'role': message.role, 'content': message.content})
+            else:
+                raise ValueError(f"role {message.role} is not supported")
```
Comment (ashwinb) [inline llama_stack/providers/adapters/safety/together/safety.py:51]: do you want to ignore such messages from a conversation or error out? erroring out is going to make this somewhat hard to use and will put the onus of filtering on the user.

File: llama_stack/providers/adapters/safety/together/safety.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+        # construct Together request
+        response = await asyncio.run(get_safety_response(together_api_key, api_messages))
```
Comment (ashwinb) [inline llama_stack/providers/adapters/safety/together/safety.py:54]: why `asyncio.run()`? you are already in an async function

File: llama_stack/providers/adapters/safety/together/safety.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    response = client.chat.completions.create(messages=messages, model="meta-llama/Meta-Llama-Guard-3-8B")
+    if len(response.choices) == 0:
+        return SafetyViolation(violation_level=ViolationLevel.INFO, user_message="safe")
+
+    response_text = response.choices[0].message.content
+    if response_text == 'safe':
+        return SafetyViolation(violation_level=ViolationLevel.INFO, user_message="safe")
+    else:
+        parts = response_text.split("\n")
+        if not len(parts) == 2:
```
Comment (ashwinb) [inline llama_stack/providers/adapters/safety/together/safety.py:68]: use `x != 2` instead of `not x == 2`?

**Review by ashwinb — CHANGES_REQUESTED** [review]
A few comments

File: llama_stack/providers/adapters/safety/together/safety.py (hunk 3 above)
Comment (yogishbaliga) [reply llama_stack/providers/adapters/safety/together/safety.py:28]: how to install pre-commit here?

Comment (ashwinb) [reply llama_stack/providers/adapters/safety/together/safety.py:28]: you can just do a `pip install pre-commit`. Then you can run `pre-commit install` which will install it as a git hook. Then when you make a commit, pre-commit will run its hook which includes linters, etc..

File: requirements.txt (hunk 1 above)
Comment (yogishbaliga) [reply requirements.txt]: done

File: llama_stack/providers/registry/safety.py (hunk 2 above)
Comment (yogishbaliga) [reply llama_stack/providers/registry/safety.py]: done

File: llama_stack/providers/adapters/safety/together/safety.py (hunk 3 above)
Comment (yogishbaliga) [reply llama_stack/providers/adapters/safety/together/safety.py]: done

File: llama_stack/providers/adapters/safety/together/safety.py (hunk 5 above)
Comment (yogishbaliga) [reply llama_stack/providers/adapters/safety/together/safety.py]: done

File: llama_stack/providers/adapters/safety/together/safety.py (hunk 6 above)
Comment (yogishbaliga) [reply llama_stack/providers/adapters/safety/together/safety.py]: done

File: llama_stack/providers/adapters/safety/together/safety.py (hunk 7 above)
Comment (yogishbaliga) [reply llama_stack/providers/adapters/safety/together/safety.py]: done

**Comment by yogishbaliga** [comment]
@ashwinb comments addressed. PTAL

File: llama_stack/providers/adapters/safety/together/safety.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+class TogetherSafetyImpl(Safety):
+    def __init__(self, config: TogetherSafetyConfig) -> None:
+        self.config = config
+
+    async def initialize(self) -> None:
+        pass
+
+    async def run_shield(
+            self, shield_type: str, messages: List[Message], params: Dict[str, Any] = None
```
Comment (ashwinb) [inline llama_stack/providers/adapters/safety/together/safety.py:28]: hey this still does not look properly formatted -- were you able to get `pre-commit` to run?

File: llama_stack/providers/registry/safety.py [hunk 9]
```diff
@@ -30,6 +31,19 @@ def available_providers() -> List[ProviderSpec]:
...
+        ),
+        remote_provider_spec(
+            api=Api.safety,
+            adapter=AdapterSpec(
+                adapter_id="together",
+                pip_packages=[
+                    "together",
+                ],
+                module="llama_stack.providers.adapters.safety.together",
+                config_class="llama_stack.providers.adapters.safety.together.TogetherSafetyConfig",
```
Comment (ashwinb) [inline llama_stack/providers/registry/safety.py:44]: you have not registered your header validator here using `provider_data_validator`

File: requirements.txt [hunk 10]
```diff
@@ -7,4 +7,4 @@ prompt-toolkit
 python-dotenv
 pydantic
 requests
-termcolor
+termcolor
```
Comment (ashwinb) [inline requirements.txt:10]: undo this change? (empty line?)

**Review by ashwinb — CHANGES_REQUESTED** [review]
Can you add a Test Plan in your summary in terms of how you tested this?

File: llama_stack/providers/adapters/safety/together/safety.py (hunk 8 above)
Comment (yogishbaliga) [reply llama_stack/providers/adapters/safety/together/safety.py:28]: yes, ran the pre-commit. Let me try it again

**Comment by yogishbaliga** [comment]
This PR is no longer required.
